docs/smart-contracts/entrypoints.md

```diff
@@ -12,77 +12,55 @@ The entrypoints in a Tezos smart contract must meet these specifications:
-For example, an entrypoint can call other entrypoints in its contract or entrypoints in other contracts.
+The only effects that an entrypoint can have are changes to its storage and new operations that are run after the entrypoint completes.
+An entrypoint can call other entrypoints in its contract or entrypoints in other contracts.
 
 ## Example entrypoints
 
-Here is a very basic example of contract with two entrypoints:
+The contract in the tutorial [Create a smart contract](../tutorials/smart-contract) has two entrypoints:
 
 | Entrypoint | Description |
 | --- | --- |
-| `add` | `add` takes an `int` as a parameter and adds it to the previous value of the storage |
-| `reset` | `reset` takes no parameter and replaces the storage with 0 |
-
-<!-- TODO
-
-## Default entrypoint behavior
-
-From Raphael:
-
-> When calling a contract, there is no difference between explicitly passing default as entrypoint name and passing no entrypoint at all.
-
-> When defining and originating a contract, the entrypoints are declared as annotations in the parameter type of the contract. There may or not be a %default in this parameter type. If there is no %default annotation, the whole parameter type is the type of the default entrypoint.
-
-> you can still have other entrypoints but they are special cases of what you can achieve by calling the default entrypoint.
-
-For example, if the parameter type is or (unit %be_nice) (unit %be_nasty), then calling the contract with:
-- Unit on the be_nice entrypoint is equivalent to
-- Left Unit on the default entrypoint
-and:
-- Unit on the be_nasty entrypoint is equivalent to
-- Right Unit on the default entrypoint.
-
-
-Each contract has a special entrypoint named "default," which runs when a caller calls the contract without specifying an entrypoint, such as if the caller sends tez to the contract.
-Senders do not pass parameters to this endpoint, though technically it accepts a parameter of the type `Unit`.
--->
-
+| `increment` | `increment` takes an `int` as a parameter and adds it to the previous value of the storage |
+| `decrement` | `decrement` takes an `int` as a parameter and subtracts it from the previous value of the storage |
 
-<!-- TODO old docs say "if this entrypoint exists" ; doesn't it always exist? -->
-<!--
-  - `default` doesn't take any parameter (or more specifically, its parameter is of type `Unit`). If this entrypoint exists, it will be executed any time the contract is called without specifying any entrypoint or parameter. This is in particular the case when a user or a contract simply sends some tez to the contract, as you do when sending tez to a user.
--->
+## The default entrypoint
 
+All contracts have a default entrypoint.
+This entrypoint runs when a caller calls the contract without specifying an entrypoint, such as when a caller sends tez to the contract.
+By default, the default entrypoint accepts the tez and does nothing.
```

Comment:
No, there is not default behavior of the default entrypoint. If you want your contract to accept tez on its default entrypoint, this is something that you need to code.

Reply:
@rafoo Can you clarify here? Is is possible to call the default entrypoint if it is not explicitly defined in the code?

Reply:
You can always call the default entry point, but you have to pass it the type of argument it expects. You may pass it tez only if it accepts this.

What is the contract you deployed?

Reply:
@NicNomadic I deployed a simple version of the Counter contract to KT1R2LTg3mQoLvHtUjo2xSi7RMBUJ1sJkDiD:

```
namespace Counter {
  type storage = int;
  type returnValue = [list<operation>, storage];

  // Increment entrypoint
  @entry
  const increment = (delta : int, store : storage) : returnValue =>
    [list([]), store + delta];

  // Decrement entrypoint
  @entry
  const decrement = (delta : int, store : storage) : returnValue => {
    return [list([]), store - delta];
  }

  // Reset entrypoint
  @entry
  const reset = (_p : unit, _s : storage) : returnValue =>
    [list([]), 0];
}
```

I assume that the default entrypoint takes a parameter of type unit (because I have not implemented it to accept any different parameter), so I try to call it with commands like these:

```
octez-client --wait none transfer 5 from tim to KT1R2LTg3mQoLvHtUjo2xSi7RMBUJ1sJkDiD --burn-cap 0.1
octez-client --wait none transfer 5 from tim to KT1R2LTg3mQoLvHtUjo2xSi7RMBUJ1sJkDiD --arg 'unit' --burn-cap 0.1
```

These commands return errors like this:

```
Invalid argument passed to contract KT1R2LTg3mQoLvHtUjo2xSi7RMBUJ1sJkDiD.
At (unshown) location 0, value Unit is invalid for type or unit (or int int).
At (unshown) location 0, invalid primitive Unit, only Left or
Right can be used here.
``` 

Seems like the error is prompting me to provide the parameter for the named entrypoints, so I don't know how to call the default entrypoint.

Reply:
If I call 

```
octez-client --wait none transfer 5 from tim to KT1R2LTg3mQoLvHtUjo2xSi7RMBUJ1sJkDiD --arg 'Left (Unit)' --burn-cap 0.1
``` 

it triggers the `reset` entrypoint, not the default entrypoint.

Reply:
The default entrypoint in the compiled Michelson contract is the one (invisible in your SmartPy program) that dispatches to the different entry points, depending on the value you pass. In your last example, it calls the reset entrypoint *because* you passed it argument `Left (Unit)`; it unwraps this argument and passes just `Unit` to reset.

So, I guess that in high-level languages such as SmartPy you may say there's no default entrypoint. And as Tezos developers usually don't program in Michelson, you may perhaps tell just this, or elude the notion altogether.

Reply:
ok, so there isn't really a default entrypoint that's separate from the entrypoints I define in my high-level language. Instead, contracts have only one entrypoint and the code branches based on the parameter value. How's this?

> The Michelson code for a smart contract does not divide its code into separate codeblocks for different entrypoints like higher-level languages do. Instead, the Michelson code for a contract has a single entrypoint known as the default entrypoint. This default entrypoint runs a branch of the contract code based on the parameters that clients pass to the contract.
> 
> For example, when you compile the contract in the tutorial [Create a smart contract](http://localhost:3000/tutorials/smart-contract) to Michelson, its first line defines the parameter type that the contract accepts:
> 
> `parameter (or (unit %reset) (or (int %decrement) (int %increment)))`
>
> To call the `reset` entrypoint, clients technically call the default entrypoint and pass the Michelson-encoded parameter `LEFT Unit`. This parameter value means that the left value of the parameter type, which is annotated `%reset`, is set to the value `Unit`, which means no value. The Michelson code uses the `IF_LEFT` command to check if the left value of the parameter is defined and runs the `reset` entrypoint code.
>
> Similarly, to call the `decrement` entrypoint and pass 5, clients call the default entrypoint and pass `Right (Left 5)`. This parameter value puts the value 5 in the integer value that is annotated `%decrement`. Then the code branches based on this parameter value.

Reply:
Yes, I assume this is the kind of code produced by any compiler to Michelson (but I didn't check this).

Note however that when you directly write Michelson code, you may deviate from this in at least two ways:
1. you are not forced to use IF_LEFT/IF_RIGHT to check your parameter and branch to a corresponding code section; for example, you can ignore the parameter and always call the same code, without any value; or you can use IF_LEFT/IF_RIGHT to compute some integer value out of it, and always call the same code with this value
2. you may annotate some branch of the parameter with the name "default": in this case, the root of the OR type may never be used, unless it is named differently!
But okay, these cases are purely theoretical, and pretty contorted, so it is out of question to describe them in the devs doc. Even, I wonder if all the "Implementation details" section makes sense for them. Or you may leave it but warn people that these are low-level details that may be skipped unless they want to known how their contracts are compiled. (I never saw a Java manual explaining what the programs look like in bytecode!)

Reply:
What about moving this section in the page for Unity, it it's only needed there? Or in a page a bit more general about web3 libraries for Tezos?

Well, I agree that it *marginally* impacts all the callers of contracts using octez-client, *if* they forget to specify an entry point. When you do indicate an entry point to octez-client, you just give the argument as that entry point expects it, no need to wrap it yourself as the default entry point would expect it, right?
If so, it's not so important why the call fails when no entry point is specified, users may just assume that they must specify an entry point, and we avoid all these hairy details.

Reply:
In other words, does this work:

    octez-client --wait none transfer 5 from tim to KT1R2LTg3mQoLvHtUjo2xSi7RMBUJ1sJkDiD --entrypoint "reset" --arg 'unit' --burn-cap 0.1

Reply:
It has to be a capital U in Unit, but yes, that works:
```
octez-client --wait none transfer 0 from tim to KT1R2LTg3mQoLvHtUjo2xSi7RMBUJ1sJkDiD --entrypoint "reset" --arg 'Unit' --burn-cap 0.1
```

Reply:
Maybe this is a good example of what's happening here: These two calls are equivalent for the contract we're discussing:

```
octez-client --wait none transfer 0 from tim to KT1R2LTg3mQoLvHtUjo2xSi7RMBUJ1sJkDiD --arg 'Left Unit' --burn-cap 0.1
```
```
octez-client --wait none transfer 0 from tim to KT1R2LTg3mQoLvHtUjo2xSi7RMBUJ1sJkDiD --entrypoint "reset" --arg 'Unit' --burn-cap 0.1
```

Reply:
yes it's a good example, independently of where you explain it

Reply:
So if you leave it here, please start the section with a note that this section may be skipped at a first reading, or so.

Reply:
I mean, both sections "Implementation details: the default entry point" and "Entrypoint implementation". BTW, I don't see well the difference between the two, perhaps they should be collapsed.

Reply:
Collapsed the sections to one section named "Implementation details: the default entrypoint."

Added this note that this info is very technical and can be ignored in most cases:

> In most cases, developers can ignore the default entrypoint and imagine that the compiled Michelson code has multiple entrypoints like the higher-level code. However, in some cases, you may need to consider how the contract actually decides which code to run and how clients trigger this code.